aclk/mqtt.c

```diff
@@ -42,7 +42,7 @@ void _link_shutdown()
     return;
 }
 
-int _link_lib_init(char *aclk_hostname, int aclk_port, void (*on_connect)(void *), void (*on_disconnect)(void *))
+int _link_lib_init(char *aclk_hostname, int aclk_port, const char *aclk_proxy, void (*on_connect)(void *), void (*on_disconnect)(void *))
```

Comment:
Many of the changes in this PR are to read the proxy string in the top layer and pass it down through the calls into the lower layer connection. Why not just read the connection in the lower layer? The common pattern to do this in the netdata code is a static variable inside the function that would use it that is initialised on first access.

Reply:
Rechecking my comments: "read the connection" should be "read the configuration".

Reply:
reason for this is already preexisting code:
```C
aclk_hostname = config_get(CONFIG_SECTION_ACLK, "agent cloud link hostname", ACLK_DEFAULT_HOST);
aclk_port = config_get_number(CONFIG_SECTION_ACLK, "agent cloud link port", ACLK_DEFAULT_PORT);
```

therefore i put also proxy reading in the same place for consistency (which as a result means you have to pass this parameters down the chain):
```C
aclk_proxy = config_get(CONFIG_SECTION_ACLK, ACLK_PROXY_CONFIG_VAR_NAME, "");
```

Reply:
I don't think that having all of the `config_` calls about a particular section in the same function is a benefit - to check all the access points to a particular section a recursive grep is needed anyway as we cannot guarantee this in general. Passing the parameters down across the layers adds unnecessary complexity. I think it is better to read it in the lower layer.

Reply:
If we want to have the config exposed in the api (even if the aclk is not active) we need to do the calls to the config parameters twice. 

Now if a config option is added to disable the cloud, we need to read that option first and make a decision if we want to hide the rest of the ACLK related configuration 
So  if (example)

```
agent cloud link functionality = no
```

we won't read the rest of the parameters or set defaults

I agree passing the parameters two or three layers down is not needed (the aclk layer needs simplification in that regard as well)

Reply:
also if it is at this level (upper) it will be common for possible HTTPS fallbacks etc.

Reply:
@stelfrag that raises an interesting question - I don't think we have config options that override other config options yet. Is that possible given the current way the configuration file is loaded?

Reply:
@underhood it would still be less complex to read the config option in the (not yet designed) HTTPS component and use it there.